### ISSUE TITLE
flann::Index does not copy features matrix

### ISSUE THREAD

**Issue by wpalfi** [opencv/opencv#12606]
##### System information (version)
- OpenCV =>3.4.1
- Operating System / Platform =>Windows
- Compiler => VS15 x64

##### Detailed description
When i create a `cv::flann::Index`, it holds a pointer to the `features` matrix data. Calls to `radiusSearch` fail when the `features `matrix is destroyed or changed from outside. `cv::flann::Index` should hold a private copy of the `features` matrix instead.

##### Steps to reproduce

```.cpp
cv::Mat1f features(1, 1, 0.f);
cv::flann::Index flann_index(features, cv::flann::KDTreeIndexParams(1));
cv::Mat indices, dists;
cv::Mat1f query(1, 1, 0.f);
EXPECT_EQ(1, flann_index.radiusSearch(query, indices, dists, 1., 32)); //ok

EXPECT_EQ(features.u->refcount, 1);
features(0) = 100;
EXPECT_EQ(1, flann_index.radiusSearch(query, indices, dists, 1., 32)); //fails
```

**Comment by berak** [comment]
it's a known pitfall.

the `flann::Index` just (shallow) copies a pointer to the train data, 
if your `features` go out of scope , before you're done with the *later prediction* (radiusSearch), 
it's nasal demons all the way down.

but again, copying it would be quite expensive, and probably not an option for large datasets.

maybe we should highlight this better in the documentation ?

**Comment by wpalfi** [comment]
@berak I can't see your comment here, why?

> copying it would be quite expensive

At least for cv::Mat it would not be expensive.

**Comment by savuor** [comment]
fixed by #14898
